core/src/main/java/com/sequenceiq/cloudbreak/ssh/SshKeyService.java

```diff
@@ -0,0 +1,72 @@
+package com.sequenceiq.cloudbreak.ssh;
+
+import java.security.KeyPair;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.certificate.PkiUtil;
+import com.sequenceiq.cloudbreak.core.bootstrap.service.ClusterDeletionBasedExitCriteriaModel;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.Cluster;
+import com.sequenceiq.cloudbreak.orchestrator.host.HostOrchestrator;
+import com.sequenceiq.cloudbreak.orchestrator.host.OrchestratorStateParams;
+import com.sequenceiq.cloudbreak.orchestrator.model.Node;
+import com.sequenceiq.cloudbreak.service.GatewayConfigService;
+import com.sequenceiq.cloudbreak.util.StackUtil;
+
+@Component
+public class SshKeyService {
+    @Inject
+    private HostOrchestrator hostOrchestrator;
+
+    @Inject
+    private GatewayConfigService gatewayConfigService;
+
+    @Inject
+    private StackUtil stackUtil;
+
+    public KeyPair generateKeyPair() {
+        return PkiUtil.generateKeypair();
+    }
+
+    public void addSshPublicKeyToHosts(Stack stack, String user, KeyPair keyPair, String authKeysComment) throws Exception {
+        OrchestratorStateParams stateParams = createSshStateParams(stack, user, keyPair, authKeysComment, "ssh.remove_ssh_publickey");
+        hostOrchestrator.runOrchestratorState(stateParams);
+        stateParams.setState("ssh.add_ssh_publickey");
+        hostOrchestrator.runOrchestratorState(stateParams);
+    }
+
+    public void removeSshPublicKeyFromHosts(Stack stack, String user, String authKeysComment) throws Exception {
+        OrchestratorStateParams stateParams = createSshStateParams(stack, user, null, authKeysComment, "ssh.remove_ssh_publickey");
+        hostOrchestrator.runOrchestratorState(stateParams);
+    }
+
+    private OrchestratorStateParams createSshStateParams(Stack stack, String user, KeyPair keyPair, String authKeysComment, String saltState) {
```

Comment:
`saltState` is always "ssh.remove_ssh_publickey"
could be moved to a constant and inlined at line #54